### ISSUE TITLE
iOS 8.x modal scroll issue

### ISSUE THREAD

**Issue by kerryusry** [twbs/bootstrap#14839]
Modal scrolling appears to be broken in iOS 8, but only in some cases. If I fill a modal with something simple like a bunch of text in paragraph tags it can handle it, but anything else like a grid stops scrolling and the background page scrolls. The same content in iOS 7 no problems. Wondering if the issue is known and if there is any kind of workaround I am stumped.

**Comment by dongepulango** [comment]
I haven't seen someone pointed out that you can't scroll pass the modal footer buttons, it's impossible to click it because bounces back when you scroll.
I think this only happens when the text/content of the modal is beyond the screen size. Only way to get out is to pinch to zoom out. I saw it on the Docs live demo.

**Comment by hugohenrique** [comment]
I created a [new version](http://github.com/hugohenrique/modal) with the fix I simplified by removing the transitions.

**Comment by intellix** [comment]
Amongst the above "fixes" that disable inertia scrolling, the body still scrolls within the background. I've found the following to fix that within iOS 8.1, Safari 8, Chrome 40, Firefox 35 on the body class:

```css
.modal-open {
    position: fixed;
    width: 100%;
    overflow: hidden;
}
```

After a bit of playing around, i've found that this within JS will cause a flash... and then work with inertia scrolling. 

```js
var modal = document.querySelector('.modal');
modal.style.position = 'initial';
setTimeout(function() {
    modal.style.position = 'fixed';
}, 50);
```

**Comment by cvrebert** [comment]
@mdo Ping.

**Comment by mdo** [comment]
Sorry, haven't been following along much here. Are folks suggesting just that we remove the overflow scrolling stuff?

**Comment by cvrebert** [comment]
@mdo That's the only remedy that's been seriously suggested so far.

**Comment by dbonicelli-oa** [comment]
Not sure if this is the most elegant solution, but adding this class to the modal body does the trick.
This makes sure the header and footer are always visible and the only thing that scrolls is the body. I personally prefer this since if you have buttons in the footer they are always visible without having to scroll.

```
.modal-body-touch {
    min-height: 100px;
    max-height: 300px;
    overflow: auto;
    -webkit-overflow-scrolling: touch;
}

@media (min-height: 600px) {
    .modal-body-touch {
        max-height: 500px;
    }
}

@media (min-height: 800px) {
    .modal-body-touch {
        max-height: 700px;
    }
}

@media (min-height: 1000px) {
    .modal-body-touch {
        max-height: 800px;
    }
}

@media (min-height: 1200px) {
    .modal-body-touch-body {
        max-height: 1000px;
    }
}
```

And then on the modal dialog:

```
<div class="modal fade">
	<div class="modal-dialog">
	    <div class="modal-content">
	        <div class="modal-header">
	        </div>
	        <div class="modal-body modal-body-touch">
	        </div>
	        <div class="modal-footer">
	        </div>
	    </div>
	</div>
</div>
```

**Comment by sergiom88** [comment]
I have a similar problem (almost the same).
I have a scrollable div with -webkit-overflow-scroll: touch; and I place an overlay on top (when the side menu opens). In iOS 8 the overlay doesnt stop the div under it from scrolling.
Any thoughts? anyone with the same problem?

**Comment by Raiden38** [comment]
Any plans to have this completly fixed in 3.3.5? I see the milestone has been changed but it's not part of the 3.3.5 list so far? Thanks!

**Comment by mdo** [comment]
The use of `-webkit-overflow-scroll: touch;` apparently doesn't have anything to do with this—I still have the same problem without that property.

**Comment by brycepj** [comment]
For me, removing the property made scrolling a modal significantly less bad. Still a little glitchy, but before it was unusable on iOS.

**Comment by mdo** [comment]
@brycepj How so? Have an example to demonstrate that?

**Comment by brycepj** [comment]
Sorry, I haven't contributed to Bootstrap and don't really know the protocol. But here are two videos I just made using an iOS simulator, with iPhone 5 (8.3). We were able to reproduce them on other iOS devices and simulators with the same version of iOS. 

The only difference was that on the second, we are overriding `touch` with `auto` on `.modal`. Selecting an input to fill seems to be what really starts the problem for us. 

[-webkit-overflow-scroll not removed (video)](https://youtu.be/qoV9kMyOXEA)

[-webkit-overflow-scroll removed (video)](https://youtu.be/86om838JUCU)

I imagine there could be complicating factors in our app, and I don't really have time to narrow this down to use case. Perhaps just use this as a data point. Wish I could be of more help! I heart Bootstrap!

**Comment by realdlee** [comment]
@brycepj @mdo Setting -webkit-overflow-scrolling to `auto` didn't resolve this for me, so we're still seeing the same thing shown in your "not removed" video.  We're using bootstrap 3.2 on this app.  Might try upgrading to the latest version in case that makes a difference.

**Comment by mdo** [comment]
@cvrebert We should get a bug report for this and close this issue out, no? Doesn't seem like something we can easily address ourselves.

**Comment by bensinclair** [comment]
Removing `-webkit-overflow-scrolling` on iOS makes it clunky and a horrible experience in general.

@intellix's suggestion worked best for me. I seem to have removed the flicker be was experiencing setting the position to  `absolute` rather than `initial`.

My CSS:

````
.modal-open {
    position: fixed;
    overflow: hidden;
}
````

My JS that I run within `$('.modal').on('shown.bs.modal', ...`:

````
var modal = document.querySelector('.modal');
modal.style.position = 'absolute';
setTimeout(function() {
    modal.style.position = 'fixed';
}, 50);
````

I tried changing the above JS to pure JQuery but found it had a flicker right to left. So at the moment using Vanilla JS is working best for me.

Can anyone else give this a try to verify it's working for you as well?

**Comment by jaketoolson** [comment]
@bensinclair your solution flickers for me.

**Comment by bensinclair** [comment]
@jaketoolson Thanks for trying. I must have some custom CSS somewhere that's fixing the flicker issue. If I can figure out what it is I'll share.

**Comment by ackernaut** [comment]
I’ve had a similar problem where a modal will not scroll on iOS. Setting this property `-webkit-overflow-scrolling: auto;` fixes it. Although I miss the bouncing scroll, it was bad UX in this case to not be able to interact with a form and buttons that were inaccessible.

I suspect that there is a bug with the overflow property and some other layout property on the page like `transform: translate` although I have not been able to confirm anything. I've seen other layout bugs before when using `transform: translate` that makes me think this.

Thanks for helpful hints, everybody!

**Comment by alexkilroy** [comment]
Any update on this. I am having the same issue.

**Comment by jcanepa** [comment]
This helped me: http://davidwalsh.name/scroll-iframes-ios

**Comment by mduveborg** [comment]
I still have this issue after using an input field..

https://jsbin.com/hagiyojufu

**Comment by xt0rted** [comment]
A similar issue happens in IE11 on Windows Phone. If the modal overflows the page will scroll once you get to the end. The fix for this was to set
```css
.modal-open {
    position: fixed;
}
```

**Comment by mduveborg** [comment]
I finally found a solution.. removing the overflow-y: scroll on .modal and having a wrapper div inside the .modal-div with:

.modal__scroll {
  max-height: 100%;
  overflow-y: auto;
  overflow-x: hidden;
}

**Comment by moodysalem** [comment]
@mduveborg , that seems to make it a lot better for me too, thanks! 

The browser can still scroll the body of the page but only after first scrolling to the bottom of the modal and then waiting for the inertia to stop, or only for a brief second after closing the virtual keyboard.

**Comment by krowa-kura** [comment]
.

17 sie 2015 14:43 "mduveborg" <notifications@github.com> napisał(a):

**Comment by JDiApice** [comment]
Hey guys so i think i found a fix.  This is working for me on iphone and android at the moment.  Its a mash up of hours upon hours of searching, reading and testing.  So if you see parts of your code in here credit goes to you lol.

@media only screen and (max-device-width:768px){

    body.modal-open {
        // block scroll for mobile;
        // causes underlying page to jump to top;
        // prevents scrolling on all screens
        overflow: hidden;
        position: fixed;
    }
}

body.viewport-lg {
    // block scroll for desktop;
    // will not jump to top;
    // will not prevent scroll on mobile
    position: absolute; 
}

body {  
  overflow-x: hidden;
   overflow-y: scroll !important;
}

The reason the media specific is on there is on a desktop i was having issues with when the modal would open all content on the page would shift from centered to left.  Looked like crap.  So this targets up to tablet size devices where you would need to scroll.   There is still a slight shift on mobile and tablet but its really not much.    Let me know if this works for you guys.  Hopefully this puts the nail in the coffin

**Comment by allisan** [comment]
Worked beautifully for me. Kudos.

**Comment by razvanphp** [comment]
@JDiApice: from my point of view, if it still jumps to the top, this issue is not fixed.

**Comment by JDiApice** [comment]
@razvanphp give it a shot first before you say its not fixed.  It worked for me on the latest version of bootstrap.  Allisan said it worked for her.  We have had over 1000+ users of our site say the scroll works across android and iphone.  It seems to have fixed our sites issue.

**Comment by razvanphp** [comment]
What I meant is that in my case, it is not acceptable to jump to the top from UX point of view, because the modal is opened from different scroll positions, and when the modal is closed, I would have to scroll back where the button that opened the modal is positioned so the user can continue reading the main page.

Sorry, I meant no disrespect, just that your solution would solve only partially the problem.

**Comment by happy2deepak** [comment]
the solution by @JDiApice creates another problem, if you have implemented Sticky footer. the footer comes up.

**Comment by TylerOrtiz** [comment]
I have tried the 
```css 
{
    -webkit-overflow-scrolling: auto;
}
```
option and that seems to work decently well for me too.

However, it does not work if you start your finger tap on an input field. This seems to be some of the root of the problem for me, if I go and start a tap on just the modal body and scroll down it will work fine, but if i start my tap on an input field element the modal will not scroll at all.

This actually occurs in both mobile Chrome and safari for me.

**Comment by rafael-paiva** [comment]
I'm having same issue in my app. The modal is fine when I first open it, but in my case, after changing the Country value and updating address fields to match international address pattern, for example, the modal won't scroll.

I tried `-webkit-overflow-scrolling: auto;` and makes it a little better but not completely decent. Like mentioned by @TylerOrtiz it seems that it doesn't work when you finger tap on a input field. I'll try other solutions proposed on this thread and give an update later.

**Comment by danazkari** [comment]
Same is happening to me on webapp, I have an input, when the user types something and then hits 'done' then it stops scrolling whatsoever, however, when the modal gets destroyed and reconstructed, then it scrolls again, that is, until I hit the input text again.

**Comment by jbhatab** [comment]
@JDiApice This helped me get to a working solution for my app. I only needed to fix the modal/scroll issue for one page on the ipad so I modified it a little.

```
@media only screen and (device-width:768px){
  body#photo_sessions_store_calendar.modal-open {
      // block scroll for mobile;
      // causes underlying page to jump to top;
      // prevents scrolling on all screens
      overflow: hidden;
      position: fixed;
  }

  body#photo_sessions_store_calendar {  
    overflow-x: hidden;
    overflow-y: scroll !important;
  }
}
```

Every ipad has the same device width of 768px and it stays 768px even if you are landscape. There's another attribute called avalWidth and availHeight that change when rotating, but the device-height stays constant. This won't work for everyone though if you need to make things work for android and phones.

**Comment by RisingSun1116** [comment]
Please do this code.
I believe you can fix the issue if you use following code.
$('body').css("position", "static");
$('body').animate({scrollTop: scrollTo}, 0);

**Comment by Quirksmode** [comment]
Has this been fixed now? It's scary to think this was opened Oct 24th last year and there still doesn't seem to be a concrete solution in place. Is the issue with iOS itself? Does this therefore affect all lightboxes or just bootstraps implementation of it?

Desperate for a fix that works, using -webkit-overflow-scrolling: auto; sadly causes the modal to jump to the top every time an input is touched

**Comment by stephanec1** [comment]
+1

**Comment by mdo** [comment]
We've no solution to offer I'm afraid. Given how far along v3 is and how close we are to killing support for it because of v4, I'm calling it. Closing as a won't fix and a browser bug :\.

**Comment by moodysalem** [comment]
The fix I've settled on is the position:fixed, width:100% on the modal-open class, plus using javascript to set the top to the top css style on the body (so scroll position does not appear to change) and then taking scrolling to that "top" when the modal is closed.

This is what that looks like with the react modal I'm using. It appears to work pretty well.

      componentWillReceiveProps: function (nextProps) {
        if (this.isMounted()) {
          var st;
          if (nextProps.open && !this.props.open) {
            st = $(window).scrollTop();
            $("body").addClass("modal-open").css("top", st * -1);
          } else if (!nextProps.open && this.props.open) {
            var body = $("body"), st = parseInt(body.css("top")) * -1;
            body.removeClass("modal-open").css("top", "");
            if (!isNaN(st)) {
              $(window).scrollTop(st);
            }
          }
        }
      },

**Comment by nadzic** [comment]
It is more than one year now and we do not have solution yet. Anyone maybe succeeded ?

**Comment by nadzic** [comment]
I have solved it. The problem is with the overflow and position. If you want that is going to work perfect in Safari (apple devices) you have to do next things:

.modal {
    overflow: initial;
    position: static;
}

So, in case if i forgot anything from there, I recommend to check if most of elements inside modal window are static.

I hope it is going to help you.

**Comment by Webtransformer** [comment]
I have tried different solutions and ended up with removing the: 
-webkit-overflow-scrolling: touch;
It even works with modal in modal.

Attention...if you use the following code then links in the modal to anchors (same page) doesnt work on some devices (iphone 4, ipad mini, ipad air):

@media only screen and (max-device-width:768px){

body.modal-open {
    // block scroll for mobile;
    // causes underlying page to jump to top;
    // prevents scrolling on all screens
    overflow: hidden;
    position: fixed;
}
}

body.viewport-lg {
// block scroll for desktop;
// will not jump to top;
// will not prevent scroll on mobile
position: absolute; 
}

body {  
overflow-x: hidden;
overflow-y: scroll !important;
}

**Comment by comapedrosa** [comment]
This worked for me:
```
.modal {
    -webkit-overflow-scrolling: auto
}
```

**Comment by Taewa** [comment]
Thanks @comapedrosa !!

**Comment by drakej** [comment]
The fix from @comapedrosa is working for me on iOS 9.x with the iPhone 6S.

**Comment by Nintex** [comment]
I fixed this for UI Kit using the following CSS. Basically a combination of some of the fixes above.

@media only screen and (max-device-width:768px) {
.uk-modal-page body {
    overflow: hidden;
    position: fixed;
    width:100%;
    }
}
.uk-notouch.uk-modal-page body {
    width:100%;
}
body {  
    overflow-x: hidden;
    overflow-y: scroll !important;
    width:100%;
    padding:0;
    margin:0;
}

This might work on Bootstrap too if you change some of the classes.

**Comment by cvrebert** [comment]
This issue seems to muddle together what might be multiple similar but separate underlying bugs, but anyway, filed a relevant WebKit/Safari bug: https://bugs.webkit.org/show_bug.cgi?id=153852

**Comment by cvrebert** [comment]
Also filed a separate bug for the "scrolling starting within a text field scrolls the `<body>` instead of the modal" problem, which I believe is what
@vigilanteweb referred to in https://github.com/twbs/bootstrap/issues/14839#issuecomment-60940738
and @brycepj referred to in https://github.com/twbs/bootstrap/issues/14839#issuecomment-102416103

WebKit bug: https://bugs.webkit.org/show_bug.cgi?id=153856

**Comment by Poolaidman** [comment]
Worked for me @comapedrosa Thank you
.modal {
    -webkit-overflow-scrolling: auto
}

**Comment by castiweb** [comment]
I solved it (I hope!) with jquery. When the modal is "visible" I've settled the HTML and BODY height as the windows screen size:
`$('body').height($(window).height());
$('html').height($(window).height());`

**Comment by happy2deepak** [comment]
@castiweb I think, its also working for me on iOS devices. Checked on iPhone 4 and iPad 2 mini.

**Comment by happy2deepak** [comment]
@castiweb not working for me! issue still persists! and it affects Sticky Footer on responsive websites.

**Comment by castiweb** [comment]
@happy2deepak I think you're right! Maybe the best way is the css rule to subscibe bootstrap declaration:
.modal-open {
    position: fixed;
}
In an ather projet this solution works fine for me.

**Comment by happy2deepak** [comment]
@castiweb yeah! Thanks. this works for me to an extent ! Though I will need to handle my Sticky Footer manually, but for time-being i can enter input values and scroll on the iOS devices

**Comment by digitalhuman** [comment]
This fixed it for me on iPhone 4s and above

```
.modal
{
    -webkit-overflow-scrolling: auto !important;
    overflow-y: auto !important;
}
```
